### PR TITLE
Allow specifying cell/gene prefix for all filtering attributes

### PR DESCRIPTION
**Pull request by nh3**
This PR allows specifying cell/gene prefix for filtering attributes in the format of "c:" or "g:". If not specified, it will look for matching columns in cell/gene table and drop ambiguous attributes. This addresses https://github.com/ebi-gene-expression-group/container-galaxy-sc-tertiary/issues/84

Additional bug fixes:
Fix bug in --subset handling
Fix bug when adding adata.var['mito']

Test updated.

### CODE REVIEW

**Comment by pinin4fjords** [comment]
Why the switch from cell -> c and gene -> g? Wouldn't it be more descriptive to leave the full words?

File: scanpy_scripts/lib/_filter.py [hunk 1]
```diff
@@ -50,41 +50,41 @@ def filter_anndata(
...
+        sc.pp.filter_cells(adata, min_counts=0)
+    if 'n_cells' not in adata.var.columns:
+        sc.pp.filter_genes(adata, min_cells=0)
+    if 'n_counts' not in adata.var.columns:
+        sc.pp.filter_genes(adata, min_counts=0)
 
     if qc_vars or pct_top:
         if not pct_top:
-            pct_top = [50]
+            pct_top = [1]
```
Comment (pinin4fjords) [inline scanpy_scripts/lib/_filter.py]: Why this change?

Comment (nh3) [reply scanpy_scripts/lib/_filter.py]: > Why the switch from cell -> c and gene -> g? Wouldn't it be more descriptive to leave the full words?

It's less typing on the command line which cli users will care (I know workflow users won't :-) ), and the prefix can be hidden by the galaxy layer so change is transparent to galaxy users.

**Comment by pinin4fjords** [comment]
So, would we e.g. set the cell filtering tool in Galaxy to specify c: to everything?

File: scanpy_scripts/lib/_filter.py [hunk 2]
```diff
@@ -163,45 +184,48 @@ def _get_filter_conditions(adata, attributes, param, category, subset):
...
-        if pt_match:
-            if name not in adata.obs.columns:
-                pct_top.append(int(pt_match['n']))
-            conditions['cell']['numerical'].append([name, vmin*100, vmax*100])
+        found, cond_cat, cond_name = _attributes_exists(
+            name, attributes, 'numerical')
+        pt_match = percent_top_pattern.match(cond_name)
+        qv_match = qc_vars_pattern.match(cond_name)
+        if found > 1:
+            logging.warning('Parameter "%s" found in both cell and gene '
```
Comment (pinin4fjords) [inline scanpy_scripts/lib/_filter.py]: Why is this just a warning? Aren't we in danger of silently dropping a filter the user thinks they're applying?

File: scanpy_scripts/lib/_filter.py (hunk 1 above)
Comment (nh3) [reply scanpy_scripts/lib/_filter.py]: > Why this change?

Good catch. Was thinking of removing the whole default value thing, as it was a work around for an old bug in scanpy, but decided to leave it but incorrectly left 1 there.

**Comment by nh3** [comment]
> So, would we e.g. set the cell filtering tool in Galaxy to specify c: to everything?

yes.

File: scanpy_scripts/lib/_filter.py (hunk 2 above)
Comment (nh3) [reply scanpy_scripts/lib/_filter.py]: If it's specific to this line, then it won't happen in galaxy as all will be prefixed. There are also other warnings for skipping variable or failing to flag MT genes, etc. In general, do you think we should just exit with error messages whenever the run isn't perfect?

Comment (pinin4fjords) [reply scanpy_scripts/lib/_filter.py]: As a user I would definitely expect that if I specify a filter, then the object I receive as a result has that filter applied (I'm not necessarily monitoring my STDERR). So if something isn't valid for filtering, then I would expect the process to die.

Comment (nh3) [reply scanpy_scripts/lib/_filter.py]: Does this apply to all non-perfect situations, e.g. fail to flag MT genes, fail to transfer certain attributes when converting format etc, or do you think it's case by case and you just want it to die in this case? There are situations when users don't necessarily know what features are available in a given object or impractical find out, so opt to just specify some "common" filters to get as close to as possible. Is that a valid reason for keeping them as warnings?

Comment (pinin4fjords) [reply scanpy_scripts/lib/_filter.py]: From a pipelines point of view I need things to be quite deterministic and predictable. I'd like to know when requested operations are not successful due to errors on my part, so I can fix them. If I specify filtering on an attribute that doesn't exist, in my opinion the process should definitely die, alerting me to my idiocy. If I'm removing mitochondrial genes I need to do a good job of flagging them and the script should fail if I haven't - I'd rather the script didn't make guesses based on 'MT-' for example (I honestly don't thing that's likely to work very reliably).

But that's just our use case of course.

Comment (pinin4fjords) [reply scanpy_scripts/lib/_filter.py]: Could there be a 'permissive' switch or something to allow operation both ways?

Comment (nh3) [reply scanpy_scripts/lib/_filter.py]: Ok, I've changed the behaviour so that it dies when filter variable is unavailable or ambiguous. But lack of MT still only gives a warning. Warning messages are there for a reason, simply ignoring them is not good practice.

Comment (pinin4fjords) [reply scanpy_scripts/lib/_filter.py]: Thanks @nh3 . Obviously I wouldn't ignore warnings interactively, but running in an automated manner they're hard to catch without some dodgy grepping of STDERR or obsessive checking of logs. Maybe some use of informative exit codes? Anyway, not important right now- thanks for your efforts.
